catalyst-gateway-crates/c509-certificate/src/c509_cert.rs

```diff
@@ -0,0 +1,282 @@
+    // Encode subject
+    tbs_cert.subject.encode(&mut encoder);
+    // Encode subject public key
+    encode_subject_public_key(
+        tbs_cert.subject_public_key_algo,
+        tbs_cert.subject_public_key,
+        &mut encoder,
+    );
+    // Encode extensions
+    encode_extensions(tbs_cert.extensions, &mut encoder);
+    // Encode issuer signature algorithm
+    encode_signature_algorithm(tbs_cert.issuer_sig_algo, &mut encoder);
+    // issuerSignatureValue
+    encode_issuer_signature_value(vec![], &mut encoder);
+
+    buffer
+}
+
+// ---------------------------------------------------
+
+#[cfg(test)]
+mod test_c509_cert {
+
+    use super::*;
+    use crate::{
+        c509_enum::AttributesRegistry,
+        cbor_encoder::{AttributeRegistryValue, RelativeDistinguishedName, StringType},
+        extensions::{
+            alt_name::{GeneralName, GeneralNamesRegistry, GeneralNamesRegistryType},
+            Extension, ExtensionRegistry, ExtensionValue,
+        },
+    };
+
+    #[test]
+    fn test_generate_c509() {
+        let issuer: Name = vec![RelativeDistinguishedName {
+            name: AttributesRegistry::CommonName,
+            value: AttributeRegistryValue {
+                str_type: StringType::Utf8String,
+                str_value: "RFC test CA".to_string(),
+            },
+        }];
+        let subject: Name = vec![RelativeDistinguishedName {
+            name: AttributesRegistry::CommonName,
+            value: AttributeRegistryValue {
+                str_type: StringType::Utf8String,
+                str_value: "01-23-45-FF-FE-67-89-AB".to_string(),
+            },
+        }];
+        let pub_key_ed25519 = vec![
+            25, 191, 68, 9, 105, 84, 205, 254, 85, 41, 186, 193, 67, 220, 59, 150, 200, 80, 86,
+            170, 48, 182, 182, 203, 12, 92, 56, 173, 112, 49, 102, 225,
+        ];
+        let extensions: Extensions = vec![Extension {
+            extension_type: ExtensionRegistry::SubjectAltName,
+            critical: true,
+            extension_value: ExtensionValue::SubjectAltName(vec![GeneralName {
+                gn_name: GeneralNamesRegistry::DNSName,
+                gn_value: GeneralNamesRegistryType::Text("example.com".to_string()),
+                critical: false,
+            }]),
+        }];
+
+        let tbs_cert = TbsCertificate {
+            certificate_type: C509CertificateType::SignedC509Cert,
+            certificate_serial_number: UnwrappedBiguint::U64Value(128269),
+            issuer,
+            validity_not_before: 1672531200,
+            validity_not_after: 1767225600,
+            subject,
+            subject_public_key_algo: PubKeyAlgoRegistry::Ed25519,
+            subject_public_key: pub_key_ed25519,
+            extensions,
+            issuer_sig_algo: SignatureAlgoRegistry::Ed25519,
+        };
+        let buffer = generate_unsigned_c509_cert(tbs_cert);
+        assert_ne!(buffer.len(), 0);
+        println!("{:?}", hex::encode(buffer));
```

Comment:
Dont we want to make an assert with some expected value ?
Or you are just checking that everything just executed well without panicking ?

Reply:
Yes, but there is not many test data spefically for Ed25519 so I don't want to assume that my encoder is right.
Hence, for now I just check whether it generates something. Will add a TODO for this.

Reply:
I see, but I think we need to add some real assertion here and would be really good to find some encoding references from other places.
Because if I am not mistaken we are implementing by our own certificate building and encoding, so it should follows to some standardised spec.